Stateless/StateMachine.cs

```diff
@@ -169,8 +178,35 @@ public void Fire<TArg0>(TriggerWithParameters<TArg0> trigger, TArg0 arg0)
             InternalFire(trigger.Trigger, arg0, arg1, arg2);
         }
         
+        /// <summary>
+        /// Queue events and then fire in order.
+        /// If only one event is queued, this behaves identically to the non-queued version.
+        /// </summary>
+        /// <param name="trigger">  The trigger. </param>
+        /// <param name="args">     A variable-length parameters list containing arguments. </param>
         void InternalFire(TTrigger trigger, params object[] args)
         {
+            _eventQueue.Enqueue(new QueuedTrigger{Trigger = trigger, Args = args});
+            if (_firing)
+                return;
+
+            try
+            {
+                _firing = true;
+                while (_eventQueue.Count != 0)
+                {
+                    var queuedEvent = _eventQueue.Dequeue();
+                    InternalFireOne(queuedEvent.Trigger, queuedEvent.Args);
+                }
+            }
+            finally
+            {
+                _firing = false;

```

Comment:
"Ensure events are fired in the same order " by events do you mean triggers? I'm not sure why there should be a queue on triggers. Isn't it by design of a state machine that it transitions from state to state, and each state evaluates whether a particular trigger makes sense? What problem does the queue solve?